src/hotspot/cpu/x86/stubGenerator_x86_64_poly.cpp

```diff
@@ -0,0 +1,1028 @@
+  //    k = (t2 & ~3) + (t2 >> 2)
+  //         Y    x4  +  Y    x1
+  //    a2:a1:a0 += k
+  //
+  // Result will be in a2:a1:a0
+  __ movq(t0, t2);
+  __ movl(a2, t2); // DWORD
+  __ andq(t0, ~3);
+  __ shrq(t2, 2);
+  __ addq(t0, t2);
+  __ andl(a2, 3); // DWORD
+
+  // a2:a1:a0 += k (kept in t0)
+  __ addq(a0, t0);
+  __ adcq(a1, 0);
+  __ adcl(a2, 0); // DWORD
+}
+
+// Convert array of 128-bit numbers in quadwords (in D0:D1) into 128-bit numbers across 44-bit limbs (in L0:L1:L2)
+// Optionally pad all the numbers (i.e. add 2^128)
+//
+//         +-------------------------+-------------------------+
+//  D0:D1  | h0 h1 g0 g1 f0 f1 e0 e1 | d0 d1 c0 c1 b0 b1 a0 a1 |
+//         +-------------------------+-------------------------+
+//         +-------------------------+
+//  L2     | h2 d2 g2 c2 f2 b2 e2 a2 |
+//         +-------------------------+
+//         +-------------------------+
+//  L1     | h1 d1 g1 c1 f1 b1 e1 a1 |
+//         +-------------------------+
+//         +-------------------------+
+//  L0     | h0 d0 g0 c0 f0 b0 e0 a0 |
+//         +-------------------------+
+//
+void StubGenerator::poly1305_limbs_avx512(
+    const XMMRegister D0, const XMMRegister D1,
+    const XMMRegister L0, const XMMRegister L1, const XMMRegister L2, bool padMSG,
+    const XMMRegister TMP, const Register rscratch)
+{
+  // Interleave blocks of data
+  __ evpunpckhqdq(TMP, D0, D1, Assembler::AVX_512bit);
+  __ evpunpcklqdq(L0, D0, D1, Assembler::AVX_512bit);
+
+  // Highest 42-bit limbs of new blocks
+  __ vpsrlq(L2, TMP, 24, Assembler::AVX_512bit);
+  if (padMSG) {
+    __ evporq(L2, L2, ExternalAddress(poly1305_pad_msg()), Assembler::AVX_512bit, rscratch); // Add 2^128 to all 8 final qwords of the message
+  }
+
+  // Middle 44-bit limbs of new blocks
+  __ vpsrlq(L1, L0, 44, Assembler::AVX_512bit);
+  __ vpsllq(TMP, TMP, 20, Assembler::AVX_512bit);
+  __ vpternlogq(L1, 0xA8, TMP, ExternalAddress(poly1305_mask44()), Assembler::AVX_512bit, rscratch); // (A OR B AND C)
+
+  // Lowest 44-bit limbs of new blocks
+  __ evpandq(L0, L0, ExternalAddress(poly1305_mask44()), Assembler::AVX_512bit, rscratch);
+}
+
+/**
+ * Copy 5×26-bit (unreduced) limbs stored at Register limbs into  a2:a1:a0 (3×64-bit limbs)
+ *
+ * a2 is optional. When only128 is set, limbs are expected to fit into 128-bits (i.e. a1:a0 such as clamped R)
+ */
+void StubGenerator::poly1305_limbs(
+    const Register limbs, const Register a0, const Register a1, const Register a2,
+    const Register t0, const Register t1)
+{
+  __ movq(a0, Address(limbs, 0));
+  __ movq(t0, Address(limbs, 8));
+  __ shlq(t0, 26);
+  __ addq(a0, t0);
+  __ movq(t0, Address(limbs, 16));
+  __ movq(t1, Address(limbs, 24));
+  __ movq(a1, t0);
+  __ shlq(t0, 52);
+  __ shrq(a1, 12);
+  __ shlq(t1, 14);
+  __ addq(a0, t0);
+  __ adcq(a1, t1);
+  __ movq(t0, Address(limbs, 32));
+  if (a2 != noreg) {
+    __ movq(a2, t0);
+    __ shrq(a2, 24);
+  }
+  __ shlq(t0, 40);
+  __ addq(a1, t0);
+  if (a2 == noreg) {
```

Comment:
Please, get rid of early return and turn the check into `if (a2 != noreg) { ... }` which guards the following code.

Reply:
done (some golang-ism slipped in.. rewiring habits again)